### PR TITLE
Osl vector

### PR DESCRIPTION
**Pull request by haggi**
Update the way compound attributes are handled.
Reorganize Environment UI tab.
Remove temporarily AOV's tab ( because not working at the moment).

### CODE REVIEW

File: module/scripts/appleseed/initialize.py [hunk 1]
```diff
@@ -70,53 +70,52 @@ def removeRendererMenu(self):
...
+            envDict['gradientHorizon'].setManage(False)
+            envDict['gradientZenit'].setManage(False)
+            envDict['environmentMap'].setManage(False)
+            envDict['latlongVeShift'].setManage(False)
+            envDict['latlongHoShift'].setManage(False)
+            
+            if envType == 0:
+                envDict['environmentColor'].setManage(True)
+            # Gradient
+            if envType == 1:
```
Comment (dictoon) [inline module/scripts/appleseed/initialize.py]: It should probably be `elif` instead of `if`.

File: module/scripts/appleseed/renderer.py [hunk 2]
```diff
@@ -384,7 +382,7 @@ def addOneTabToGlobalsWindow(self, renderer, tabLabel, createProc):
             pm.setUITemplate('attributeEditorTemplate', pushTemplate=True)
             pm.formLayout(tabName)
             createProcs = ['createMayaSoftwareCommonGlobalsTab', 'createMayaSoftwareGlobalsTab']
-            for renderer in ["Corona", "arnold", "appleseed", "Thea", "Indigo", "Fuji", "Lux"]:
+            for renderer in ["arnold", "Appleseed"]:
```
Comment (dictoon) [inline module/scripts/appleseed/renderer.py]: Shouldn't appleseed be the only renderer left here?

File: src/utilities/attrtools.cpp [hunk 3]
```diff
@@ -291,6 +291,23 @@ int getEnumInt(MPlug plug)
     return value;
 }
 
+MString getEnumString(const MPlug& plug)
+{
+    MDGContext ctx = MDGContext::fsNormal;
+    MStatus stat = MS::kSuccess;
+    bool result = false;
```
Comment (dictoon) [inline src/utilities/attrtools.cpp]: `result` is unused.

File: src/utilities/attrtools.cpp [hunk 4]
```diff
@@ -456,9 +473,19 @@ MColor getColorAttr(const char *plugName, const MFnDependencyNode& dn)
 
 MVector getVectorAttr(const char *plugName, const MFnDependencyNode& dn)
 {
-    MVector c(0, 0, 0);
-    getVector(MString(plugName), dn, c);
-    return c;
+    MVector v(0, 0, 0);
+    MDGContext ctx = MDGContext::fsNormal;
+    MStatus stat = MS::kSuccess;
+    bool result = false;
```
Comment (dictoon) [inline src/utilities/attrtools.cpp]: `result` is unused.

File: src/utilities/attrtools.cpp [hunk 5]
```diff
@@ -578,6 +605,19 @@ MMatrix getMatrix(const char *plugName, const MFnDependencyNode& dn)
     return m;
 }
 
+MMatrix getMatrix(const MPlug& plug)
+{
+    MMatrix m;
+    m.setToIdentity();
+    MStatus stat;
```
Comment (dictoon) [inline src/utilities/attrtools.cpp]: `state` is unused.

File: src/utilities/attrtools.cpp [hunk 6]
```diff
@@ -578,6 +605,19 @@ MMatrix getMatrix(const char *plugName, const MFnDependencyNode& dn)
...
+    MObject matrixObject;
+    plug.getValue(matrixObject);
+    MFnMatrixData fnMat(matrixObject);
+    MMatrix mat = fnMat.matrix();
```
Comment (dictoon) [inline src/utilities/attrtools.cpp]: The next three lines can be simplified to

    return fnMat.matrix();

File: src/utilities/oslutils.cpp [hunk 7]
```diff
@@ -319,87 +319,121 @@ void OSLUtilClass::listProjectionHistory(const MObject& mobject)
     }
 }
 
+namespace {
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: Place curly brace on its own line.

File: src/utilities/oslutils.cpp [hunk 8]
```diff
@@ -319,87 +319,121 @@ void OSLUtilClass::listProjectionHistory(const MObject& mobject)
...
-    // for a color array we have in a osl shader color0 color1 color2 ... at the moment we are limited to
-    // several entries this way. I set it to ARRAY_MAX_ENTRIES.
-    MPlug plug = depFn.findPlug(sa.name.c_str());
-    if (plug.isArray())
+
+    /*
+        Arrays in OSL cannot be connected per element. e.g. a connection nodeA.outColor -> nodeB.colorList[3] is invalid.
+        Our workaround is simply to use a limited number of entries (e.g. color0, color1, color2...) and map the maya array 
+        element to the OSL entry, e.g. nodeA.outColor -> nodeB.color3.
+        Unfortunatly there is no direct relationship from the OSL attribute name with the plug name. e.g. an attribute called color1
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: *Unfortunately*

File: src/utilities/oslutils.cpp [hunk 9]
```diff
@@ -319,87 +319,121 @@ void OSLUtilClass::listProjectionHistory(const MObject& mobject)
...
-    // if we have an array plug we have do deal with several plugs.
-    // OSL is not able to connect nodes to array indices so we always have a list of parameters, e.g.
-    // for a color array we have in a osl shader color0 color1 color2 ... at the moment we are limited to
-    // several entries this way. I set it to ARRAY_MAX_ENTRIES.
-    MPlug plug = depFn.findPlug(sa.name.c_str());
-    if (plug.isArray())
+
+    /*
+        Arrays in OSL cannot be connected per element. e.g. a connection nodeA.outColor -> nodeB.colorList[3] is invalid.
+        Our workaround is simply to use a limited number of entries (e.g. color0, color1, color2...) and map the maya array
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: *Maya*

File: src/utilities/oslutils.cpp [hunk 10]
```diff
@@ -319,87 +319,121 @@ void OSLUtilClass::listProjectionHistory(const MObject& mobject)
     }
 }
 
+namespace {
+    unsigned int getArrayIndex(const std::string& value)
+    {
+        char intval[12];
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: This code is kind of fragile and dangerous. Where does the 12 come from? What hapens if `value` contains more than 11 digits?

Here is a simpler, faster and more robust implementation:

    unsigned int getArrayIndex(const std::string& value)
    {
        unsigned int result = 0;
        for (size_t i = 0, e = value.size(); i < e; ++i)
        {
            if (value[i] >= '0' && value[i] <= '9')
                result = result * 10 + value[i] - '0';
        }
        return result;
    }

File: src/utilities/oslutils.cpp [hunk 11]
```diff
@@ -319,87 +319,121 @@ void OSLUtilClass::listProjectionHistory(const MObject& mobject)
...
+        for (size_t i = 0; i < value.size(); i++)
+        {
+            if ((value[i] >= '0') && (value[i] <= '9'))
+                intval[counter++] = value[i];
+        }
+        intval[counter] = '\0';
+        return MString(intval).asUnsigned();
+    }
+
+    MString removeIndexFromName(const MString& value, const int& index)
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: Integers are small enough to be passed by value: `const int index`.

File: src/utilities/oslutils.cpp [hunk 12]
```diff
@@ -319,87 +319,121 @@ void OSLUtilClass::listProjectionHistory(const MObject& mobject)
...
+    {
+        plug = depFn.findPlug(sa.name.c_str());
+        if (plug.isNull())
+            return;
     }
 
     if (sa.type == "string")
     {
-        // in OSL we don't have option menus, they are defined in our definition by string metadatas
+        // in OSL we don't have option menus, they are defined by string metadatas
```
Comment (dictoon) [inline src/utilities/oslutils.cpp]: Start full-line comment with a capital.
Metadata is already plural.

File: src/utilities/oslutils.cpp (hunk 10 above)
Comment (haggi) [inline src/utilities/oslutils.cpp]: Great, I hoped you came up with something like that :)

**Comment by dictoon** [comment]
Thanks, looks good, merging now! You can address to two additional remarks I made in a later commit.

**Pull request merged**